docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+:toc: macro
+
+= RFC 1: T Staking Contract
+
+:icons: font
+:numbered:
+toc::[]
+
+== Proposal
+
+=== Goal
+
+The goal of this proposal is to specify a simple and secure stake delegation
+mechanism. It should enable T owners to have their wallets offline and their
+stake operated by operators on their behalf. All off-chain client software
+should be able to run without exposing operator’s private key and should not
+require any owner’s keys at all. The stake delegation should also optimize the
+network throughput without compromising the security of the owners’ stake.
+
+This proposal aims at implementing a minimum viable staking contract version
+allowing to support legacy KEEP and NU stakes, as well as native T delegations.
+The functionality of the staking contract can be further extended by the
+upgradeability of the contract code.
+
+=== Terminology
+
+Owner:: An address owning T tokens and tokens that could be converted to
+T: KEEP, NU, including grants. An owner is the ultimate holder of the tokens.
+Before stake delegation, the owner has full control over the tokens, and the
+tokens are returned to the owner after stake delegation has finished.
+Owner’s participation is not required in the day-to-day operations on the
+stake, so cold storage can be accommodated to the maximum extent.
+
+Operator:: An address of a party authorized to operate in the network on behalf
+of a given owner. The operator handles the everyday operations on the delegated
+stake without actually owning the staked tokens. An operator can not simply
+transfer away delegated tokens, however, it should be noted that the operator’s
+misbehavior may result in slashing tokens and thus the entire staked amount is
+indeed at stake.
+
+Beneficiary:: An address where the rewards for participation are sent, earned by
+the operator, on behalf of the owner. A beneficiary doesn’t sign or publish any
+protocol-relevant transactions, but any currency or tokens earned by the
+operator will be transferred to the beneficiary.
+
+Authorizer:: An address appointed by an owner to authorize applications on
+behalf of them. An application must be approved by the authorizer before the
+operator is eligible to participate.
+
+Delegated stake:: An owner’s staked tokens, delegated to the operator by the
+owner. Delegation enables KEEP owners to have their wallets offline and their
+stake operated by operators on their behalf.
+
+== Specification
+
+=== Functionality
+
+==== Delegating a stake
+
+Tokens are delegated by the owner. During the delegation, the owner needs to
+appoint an operator, beneficiary, and authorizer. The owner may decide to
+delegate just a portion of their tokens. The owner may delegate multiple times
+to different operators. For simplicity, the operator address can not be reused
```

Comment:
The proposition is that operator needs to be unique, the rest (owner, beneficiary, authorizer) can be reused between applications. Here are some examples we observe in practice on Keep mainnet that could help to understand the relationships:
- One owner delegates some part of their stake to one operator and some part of their stake to another operator; authorizer and beneficiary are controlled by the owner and reused in these two delegations.
- One owner delegates some part of their stake to one operator and some part of their stake to another operator; authorizer and beneficiary are controlled by operators so different addresses are used in these two delegations.
- Multiple owners delegate to a staking provider who authorizes on behalf of owners; each owner has their own operator address and beneficiary address but the same authorizer address is used.

> What's the relation between the owner, operator, beneficiary, and authorizer, in a model where an owner has delegated to multiple operators? Isn't it effectively separate stakes?

Correct, these are separate stake delegations. Owner/beneficiary/authorizer can be reused between delegations, operator can not.

> On top of this, how do these roles interact with the model that we agreed on with multiple allocations? Specifically, is operator a stake-level role, or is it more an application-level role? 

It is a stake delegation-level role. One operator can be authorized to operate in multiple applications.

> As far as I understand, if an owner can have multiple operators and allocations, the most natural implications is that each allocation is tied to an operator (and viceversa).

One operator can have multiple allocations. I think the "allocation" term is confusing -  it is not really an allocation but an authorization to use, and slash, up to the given amount of delegated stake. 

If we do allocations, in the strict understanding, applications generating more income would be preferred and other applications - still generating income but not that high compared to others - would be starved.

Reply:
> One owner delegates some part of their stake to one operator and some part of their stake to another operator; authorizer and beneficiary are controlled by operators so different addresses are used in these two delegations.

You lost me here. "authorizer and beneficiary are controlled by operators"? I thought these roles are always controlled by owners. Overall, I still don't get how the multiple operators per stake works in a multi-app setting.

> One operator can have multiple allocations. I think the "allocation" term is confusing - it is not really an allocation but an authorization to use, and slash, up to the given amount of delegated stake.
If we do allocations, in the strict understanding, applications generating more income would be preferred and other applications - still generating income but not that high compared to others - would be starved.

OK, I understand that the term "allocations" could imply exclusive amounts to someone, and this is not the case. Let's go with "authorizations" 👍

Reply:
> You lost me here. "authorizer and beneficiary are controlled by operators"? I thought these roles are always controlled by owners. Overall, I still don't get how the multiple operators per stake works in a multi-app setting.

There would not be multiple operators per stake. The same _owner_ can stake tokens at multiple operators, where each operator is an independent stake for all purposes. Stakes are uniquely associated with the operator addresses, and the staking contract doesn't care about the owner address except when returning the staked tokens.

The authoriser and beneficiary are roles that can be freely set by the owner when staking tokens. The authoriser and beneficiary can be controlled by either the owner or the operator, the staking contract doesn't care about the off-chain arrangements.

[Here is the original explanation of the roles](https://github.com/keep-network/keep-core/blob/main/docs/random-beacon/staking/index.adoc#roles)

Reply:
Just like @eth-r explained. I think one more thing is worth adding here as I suspect this could be confusing: operator != staking provider. An operator is just an address, same beneficiary, authorizer, and owner. From the perspective of the staking contract, who controls them does not matter. Practically, we've seen different setups: individuals controlling all 4 roles, owners delegating to staking providers but still controlling beneficiary and authorizer, owners delegating to staking providers and using beneficiary and authorizer controlled by them.